compiler/src/dotty/tools/dotc/transform/init/Objects.scala

```diff
@@ -0,0 +1,1419 @@
+   *
+   * @param thisV        The value for `this` where the variable is used.
+   * @param sym          The symbol of the variable.
+   */
+  def readLocal(thisV: Value, sym: Symbol): Contextual[Value] = log("reading local " + sym.show, printer, (_: Value).show) {
+    def isByNameParam(sym: Symbol) = sym.is(Flags.Param) && sym.info.isInstanceOf[ExprType]
+    Env.resolveEnv(sym.enclosingMethod, thisV, summon[Env.Data]) match
+    case Some(thisV -> env) =>
+      if sym.is(Flags.Mutable) then
+        // Assume forward reference check is doing a good job
+        given Env.Data = env
+        val addr = Env.varAddr(sym)
+        if addr.owner == State.currentObject then
+          Heap.read(addr)
+        else
+          errorReadOtherStaticObject(State.currentObject, addr.owner)
+          Bottom
+        end if
+      else if sym.isPatternBound then
+        // TODO: handle patterns
+        Cold
+      else
+        given Env.Data = env
+        try
+          // Assume forward reference check is doing a good job
+          val value = Env.valValue(sym)
+          if isByNameParam(sym) then
+            value match
+            case fun: Fun =>
+              given Env.Data = fun.env
+              eval(fun.code, fun.thisV, fun.klass)
+            case Cold =>
+              report.warning("Calling cold by-name alias. Call trace: \n" + Trace.show, Trace.position)
+              Bottom
+            case _: RefSet | _: Ref =>
+              report.warning("[Internal error] Unexpected by-name value " + value.show  + ". Calling trace:\n" + Trace.show, Trace.position)
+              Bottom
+          else
+            value
+
+        catch ex =>
+          report.warning("[Internal error] Not found " + sym.show + "\nenv = " + env.show + ". Calling trace:\n" + Trace.show, Trace.position)
+          Bottom
+
+    case _ =>
+      if isByNameParam(sym) then
+        report.warning("Calling cold by-name alias. Call trace: \n" + Trace.show, Trace.position)
+        Bottom
+      else
+        Cold
+  }
+
+  /** Handle local variable assignmenbt, `x = e`.
+   *
+   * @param thisV        The value for `this` where the assignment locates.
+   * @param sym          The symbol of the variable.
+   * @param value        The value of the rhs of the assignment.
+   */
+  def writeLocal(thisV: Value, sym: Symbol, value: Value): Contextual[Value] = log("write local " + sym.show + " with " + value.show, printer, (_: Value).show) {
+
+    assert(sym.is(Flags.Mutable), "Writing to immutable variable " + sym.show)
+    Env.resolveEnv(sym.enclosingMethod, thisV, summon[Env.Data]) match
+    case Some(thisV -> env) =>
+      given Env.Data = env
+      val addr = Env.varAddr(sym)
+      if addr.owner != State.currentObject then
+        errorMutateOtherStaticObject(State.currentObject, addr.owner)
+      else
+        Heap.write(addr, value)
+
+    case _ =>
+      report.warning("Assigning to variables in outer scope", Trace.position)
+
+    Bottom
+  }
+
```

Comment:
Same question as `assign`, does it need to return a value?

Reply:
Assignment returns a Unit value, e.g., `val a = (x.f = e)`